src/python/gennylib/auto_tasks.py

```diff
@@ -0,0 +1,361 @@
+import enum
+import glob
+import os
+import re
+import subprocess
+import sys
+from typing import NamedTuple, List, Optional, Set
+import pathlib
+
+import yaml
+from shrub.command import CommandDefinition
+from shrub.config import Configuration
+from shrub.variant import TaskSpec
+
+
+def _check_output(cwd, *args, **kwargs):
+    old_cwd = os.getcwd()
+    try:
+        if not os.path.exists(cwd):
+            raise Exception(f"Cannot chdir to {cwd} from cwd={os.getcwd()}")
+        os.chdir(cwd)
+        out = subprocess.check_output(*args, **kwargs)
+    except subprocess.CalledProcessError as e:
+        print(e.output, file=sys.stderr)
+        raise e
+    finally:
+        os.chdir(old_cwd)
+
+    if out.decode() == "":
+        return []
+    return out.decode().strip().split("\n")
+
+
+def _yaml_load(files: List[str]) -> dict:
+    out = dict()
+    for file in files:
+        if not os.path.exists(file):
+            continue
+        basename = os.path.basename(file).split(".yml")[0]
+        with open(file) as contents:
+            out[basename] = yaml.safe_load(contents)
+    return out
+
+
+class DirectoryStructure:
+    def __init__(self, repo_root: str):
+        self.repo_root = repo_root
+
+    def all_workload_files(self) -> Set[str]:
+        pattern = os.path.join(self.repo_root, "src", "workloads", "*", "*.yml")
+        return {*glob.glob(pattern)}
+
+    def modified_workload_files(self) -> Set[str]:
+        command = (
+            "git diff --name-only --diff-filter=AMR "
+            # TODO: don't use rtimmons/
+            "$(git merge-base HEAD rtimmons/master) -- src/workloads/"
+        )
+        print(f"Command: {command}")
+        lines = _check_output(self.repo_root, command, shell=True)
+        return {os.path.join(self.repo_root, line) for line in lines if line.endswith(".yml")}
+
+
+class OpName(enum.Enum):
+    ALL_TASKS = object()
+    VARIANT_TASKS = object()
+    PATCH_TASKS = object()
+
+
+class CLIOperation(NamedTuple):
+    mode: OpName
+    variant: Optional[str]
+    is_legacy: bool
+    output_file_suffix: str
+
+    @property
+    def repo_root(self) -> str:
+        if self.is_legacy:
+            if self.mode != OpName.VARIANT_TASKS:
+                return "./genny/genny"
+            return "../src/genny/genny"
+        return "./src/genny"
+
+    @property
+    def output_file(self) -> str:
+        if self.is_legacy:
+            return os.path.join(self.repo_root, self.output_file_suffix)
+        return self.output_file_suffix
+
+    @staticmethod
+    def parse(argv: List[str]) -> "CLIOperation":
```

Comment:
argparse?

Reply:
Debated it, but new-world will not have arguments just a single argv subcommand. Debatable if argparse is worth it but punting that for now.